fsw/cfe-core/src/inc/cfe_msg.h

```diff
@@ -0,0 +1,220 @@
+**          reporting the error.
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a checksum field, then this routine will always return \c true.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**
+** \return Boolean checksum result
+** \retval true  The checksum field in the packet is valid.
+** \retval false The checksum field in the packet is not valid or the message type is wrong.
+**/
+bool CFE_MSG_ValidateChecksum(CFE_SB_MsgPtr_t MsgPtr);
+/**@}*/
+
+/** @defgropu CFEAPIMSGCmdCode cFE Message Command Code APIs
+ * @{
+ */
+
+/*****************************************************************************/
+/**
+** \brief Sets the command code field in a software bus message.
+**
+** \par Description
+**          This routine sets the command code of a software bus message (if SB
+**          messages are implemented as CCSDS packets, this will be the function code).
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a command code field, then this routine will do nothing to
+**            the message contents and will return #CFE_SB_WRONG_MSG_TYPE.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**
+** \param[in]  CmdCode     The command code to include in the message.
+**
+** \return Execution status, see \ref CFEReturnCodes
+** \retval #CFE_SUCCESS           \copybrief CFE_SUCCESS
+** \retval #CFE_SB_WRONG_MSG_TYPE \copybrief CFE_SB_WRONG_MSG_TYPE
+**
+**/
+int32 CFE_MSG_SetCmdCode(CFE_SB_MsgPtr_t MsgPtr,
+                         uint16 CmdCode);
+
+/*****************************************************************************/
+/**
+** \brief Gets the command code field from a software bus message.
+**
+** \par Description
+**          This routine gets the command code from a software bus message (if
+**          SB messages are implemented as CCSDS packets, this will be the function
+**          code).
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a command code field, then this routine will return a zero.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**
+** \return The command code included in the software bus message header (if present).
+**         Otherwise, returns a command code value of zero.
+**/
+uint16 CFE_MSG_GetCmdCode(CFE_SB_MsgPtr_t MsgPtr);
```

Comment:
Another gripe about the API - The "SetCmdCode" above returns an `int32` status and thereby offers a distinct return code `CFE_SB_WRONG_MSG_TYPE` if you pass it a message that isn't a command.  But "GetCmdCode" is not a mirror image of this and directly return a value.  Why?

It is clear that the existing CFE_SB API probably "grew" organically by different devs at different times which can explain the differences/mismatches.  But if we are making a new API we should address these.

Again - this would be fine if this `cfe_msg.h` were header were inside the library or otherwise private/restricted to SB, but this PR is making it publicly available in `fsw/cfe-core/src/inc` ... 

Public APIs will have to be lived with and maintained for years to come, so its best to really think about them and design them.  A new message abstraction offers the chance to fix mistakes of the past, we shouldn't just propagate the old (broken/mismatched) API paradigms to a new name without addressing the shortcomings.

Recommendation for the near term is to keep this include file quasi-private by making the user explicitly add an `include_directories` to their CMake recipe to get it and/or use the `target_include_directories()` cmake function with the PUBLIC or INTERFACE type to accomplish this.  Therefore users that want to call these functions in the near term still can, but we can do a proper design cycle/review before adding new public APIs that we will be committed to.

Reply:
Also I've always wanted to typedef return status instead of just relying on int32/uint32, would now be a good time to introduce such a change?

Reply:
> Also I've always wanted to typedef return status instead of just relying on int32/uint32, would now be a good time to introduce such a change?

Lets consider such a change as a separate issue/PR.

Reply:
All great suggestions, but in the short term just moved cfe_msg.h internal to the msg module in b5c8be6.  I will do a little bit of fix-up, but there will be plenty of opportunities for further improvement.